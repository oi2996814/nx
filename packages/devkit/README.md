<p style="text-align: center;">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://raw.githubusercontent.com/nrwl/nx/master/images/nx-dark.svg">
    <img alt="Nx - Smart Monorepos · Fast CI" src="https://raw.githubusercontent.com/nrwl/nx/master/images/nx-light.svg" width="100%">
  </picture>
</p>

{{links}}

<hr>

# Nx: Smart Monorepos · Fast CI

Nx is a build system, optimized for monorepos, with plugins for popular frameworks and tools and advanced CI capabilities including caching and distribution.

This package contains a set of utilities for creating Nx plugins.

{{content}}
