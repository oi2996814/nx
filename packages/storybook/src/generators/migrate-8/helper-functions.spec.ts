import {
  addProjectConfiguration,
  getPackageManagerCommand,
  output,
  ProjectConfiguration,
  Tree,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import * as allProjects from './test-configs/all-projects.json';
import {
  getAllStorybookInfo,
  logResult,
  onlyShowGuide,
} from './helper-functions';

describe('Helper functions for the Storybook 8 migration generator', () => {
  let tree: Tree;

  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
    addAllProjectsToWorkspace(tree);
  });

  describe('getAllStorybookInfo and onlyShowGuide', () => {
    let allStorybookInfo;
    beforeEach(() => {
      allStorybookInfo = getAllStorybookInfo(tree);
    });
    it('should return all info for all projects with Storybook', () => {
      expect(allStorybookInfo).toMatchSnapshot();
    });

    it('should onlyShowGuide and the correct instructions', () => {
      const outputSpy = jest.spyOn(output, 'log').mockImplementation();
      onlyShowGuide(allStorybookInfo);
      const pm = getPackageManagerCommand();
      expect(outputSpy).toHaveBeenCalledWith(
        expect.objectContaining({
          bodyLines: [
            'You can run the following commands manually to upgrade your Storybook projects to Storybook 8:',
            '',
            '1. Call the Storybook upgrade script:',
            `${pm.exec} storybook@latest upgrade`,
            '',
            '2. Call the Storybook automigrate scripts:',
            'Run the following commands for each Storybook project:',
            `${pm.exec} storybook@latest automigrate --config-dir apps/nextapp/.storybook`,
            `${pm.exec} storybook@latest automigrate --config-dir apps/nextapp-ts/.storybook`,
            `${pm.exec} storybook@latest automigrate --config-dir apps/ngapp/.storybook`,
            `${pm.exec} storybook@latest automigrate --config-dir apps/ngapp-ts/.storybook`,
            `${pm.exec} storybook@latest automigrate --config-dir apps/rv1/.storybook`,
            `${pm.exec} storybook@latest automigrate --config-dir apps/rv2-ts/.storybook`,
            `${pm.exec} storybook@latest automigrate --config-dir apps/rw1/.storybook`,
            `${pm.exec} storybook@latest automigrate --config-dir apps/wv1/.storybook`,
            `${pm.exec} storybook@latest automigrate --config-dir apps/ww1/.storybook`,
            `${pm.exec} storybook@latest automigrate --config-dir plugin-apps/plugin/.storybook`,
            '',
          ],
          title: 'Storybook 8 Migration Guide',
        })
      );
    });
  });

  describe('logResult', () => {
    it('should create the summary file with the correct content', () => {
      logResult(tree, {
        successfulProjects: {
          nextapp:
            'npx storybook@latest automigrate --config-dir apps/nextapp/.storybook',
          rv1: 'npx storybook@latest automigrate --config-dir apps/rv1/.storybook',
        },
        failedProjects: {
          'rv2-ts': `npx storybook@latest automigrate --config-dir apps/rv2-ts/.storybook`,
          rw1: 'npx storybook@latest automigrate --config-dir apps/rw1/.storybook',
        },
      });
      expect(
        tree.read('storybook-migration-summary.md', 'utf-8')
      ).toMatchSnapshot();
    });
  });
});

export function addAllProjectsToWorkspace(tree: Tree) {
  for (const [name, project] of Object.entries(allProjects)) {
    addProjectConfiguration(tree, name, project as ProjectConfiguration);
  }
  writeMainJs(tree);
  writeViteConfig(tree);

  addPluginProjects(tree);
}

function writeViteConfig(tree: Tree) {
  tree.write(
    `apps/rv1/vite.config.js`,
    `const { defineConfig } = require('vite');`
  );

  tree.write(
    `apps/rv2-ts/vite.config.ts`,
    `const { defineConfig } = require('vite');`
  );

  tree.write(
    `apps/wv1/vite.config.ts`,
    `const { defineConfig } = require('vite');`
  );
}

function writeMainJs(tree: Tree) {
  tree.write(
    `apps/rv1/.storybook/main.js`,
    `
    const { mergeConfig } = require('vite');
    const viteTsConfigPaths = require('vite-tsconfig-paths').default;

    module.exports = {
      core: { builder: '@storybook/builder-vite' },
      stories: [
        '../src/app/**/*.stories.mdx',
        '../src/app/**/*.stories.@(js|jsx|ts|tsx)',
      ],
      addons: ['@storybook/addon-essentials'],
      async viteFinal(config, { configType }) {
        return mergeConfig(config, {
          plugins: [
            viteTsConfigPaths({
              root: '../../../',
            }),
          ],
        });
      },
    };
`
  );

  tree.write(
    `apps/rv2-ts/.storybook/main.ts`,
    `
    import type { StorybookConfig } from '@storybook/core-common';

    import { mergeConfig } from 'vite';
    import viteTsConfigPaths from 'vite-tsconfig-paths';

    const config: StorybookConfig = {
      core: { builder: '@storybook/builder-vite' },
      stories: [
        '../src/app/**/*.stories.mdx',
        '../src/app/**/*.stories.@(js|jsx|ts|tsx)',
      ],
      addons: ['@storybook/addon-essentials'],
      async viteFinal(config: any) {
        return mergeConfig(config, {
          plugins: [
            viteTsConfigPaths({
              root: '../../../',
            }),
          ],
        });
      },
    } as StorybookConfig;

    module.exports = config;
  `
  );

  tree.write(
    `apps/nextapp/.storybook/main.js`,
    `
    const path = require('path');

    module.exports = {
      core: { builder: 'webpack5' },
      stories: [
        '../components/**/*.stories.mdx',
        '../components/**/*.stories.@(js|jsx|ts|tsx)',
      ],
      addons: [
        '@storybook/addon-essentials',
        '@nx/react/plugins/storybook',
        'storybook-addon-swc',
        {
          name: 'storybook-addon-next',
          options: {
            nextConfigPath: path.resolve(__dirname, '../next.config.js'),
          },
        },
      ],
    };
  `
  );

  tree.write(
    `apps/nextapp-ts/.storybook/main.ts`,
    `
    import type { StorybookConfig } from '@storybook/core-common';
    import path from 'path';

    const config: StorybookConfig = {
      core: { builder: 'webpack5' },
      stories: [
        '../components/**/*.stories.mdx',
        '../components/**/*.stories.@(js|jsx|ts|tsx)',
      ],
      addons: [
        '@storybook/addon-essentials',
        '@nx/react/plugins/storybook',
        'storybook-addon-swc',
        {
          name: 'storybook-addon-next',
          options: {
            nextConfigPath: path.resolve(__dirname, '../next.config.js'),
          },
        },
      ],
    } as StorybookConfig;

    module.exports = config;
  `
  );

  tree.write(
    `apps/rw1/.storybook/main.js`,
    `
    const path = require('path');

    module.exports = {
      core: { builder: 'webpack5' },
      stories: [
        '../components/**/*.stories.mdx',
        '../components/**/*.stories.@(js|jsx|ts|tsx)',
      ],
      addons: [
        '@storybook/addon-essentials',
        '@nx/react/plugins/storybook',
        'storybook-addon-swc',
        {
          name: 'storybook-addon-next',
          options: {
            nextConfigPath: path.resolve(__dirname, '../next.config.js'),
          },
        },
      ],
    };
  `
  );
  tree.write(
    `apps/wv1/.storybook/main.js`,
    `
    const path = require('path');

    module.exports = {
      core: { builder: 'webpack5' },
      stories: [
        '../components/**/*.stories.mdx',
        '../components/**/*.stories.@(js|jsx|ts|tsx)',
      ],
      addons: [
        '@storybook/addon-essentials',
        '@nx/react/plugins/storybook',
        'storybook-addon-swc',
        {
          name: 'storybook-addon-next',
          options: {
            nextConfigPath: path.resolve(__dirname, '../next.config.js'),
          },
        },
      ],
    };
  `
  );
  tree.write(
    `apps/ww1/.storybook/main.js`,
    `
    const path = require('path');

    module.exports = {
      core: { builder: 'webpack5' },
      stories: [
        '../components/**/*.stories.mdx',
        '../components/**/*.stories.@(js|jsx|ts|tsx)',
      ],
      addons: [
        '@storybook/addon-essentials',
        '@nx/react/plugins/storybook',
        'storybook-addon-swc',
        {
          name: 'storybook-addon-next',
          options: {
            nextConfigPath: path.resolve(__dirname, '../next.config.js'),
          },
        },
      ],
    };
  `
  );
  tree.write(
    `apps/ngapp/.storybook/main.js`,
    `
    const path = require('path');

    module.exports = {
      core: { builder: 'webpack5' },
      stories: [
        '../components/**/*.stories.mdx',
        '../components/**/*.stories.@(js|jsx|ts|tsx)',
      ],
      addons: [
        '@storybook/addon-essentials',
        '@nx/react/plugins/storybook',
        'storybook-addon-swc',
        {
          name: 'storybook-addon-next',
          options: {
            nextConfigPath: path.resolve(__dirname, '../next.config.js'),
          },
        },
      ],
    };
  `
  );

  tree.write(
    `apps/ngapp-ts/.storybook/main.ts`,
    `
    const path = require('path');

    module.exports = {
      core: { builder: 'webpack5' },
      stories: [
        '../components/**/*.stories.mdx',
        '../components/**/*.stories.@(js|jsx|ts|tsx)',
      ],
      addons: [
        '@storybook/addon-essentials',
        '@nx/react/plugins/storybook',
        'storybook-addon-swc',
        {
          name: 'storybook-addon-next',
          options: {
            nextConfigPath: path.resolve(__dirname, '../next.config.js'),
          },
        },
      ],
    };
  `
  );
}

function addPluginProjects(tree: Tree) {
  tree.write(
    `plugin-apps/plugin/.storybook/main.ts`,
    `
    import type { StorybookConfig } from '@storybook/core-common';
    import path from 'path';

    const config: StorybookConfig = {
      core: { builder: 'webpack5' },
      stories: [
        '../components/**/*.stories.mdx',
        '../components/**/*.stories.@(js|jsx|ts|tsx)',
      ],
      addons: [
        '@storybook/addon-essentials',
        '@nx/react/plugins/storybook',
        'storybook-addon-swc',
        {
          name: 'storybook-addon-next',
          options: {
            nextConfigPath: path.resolve(__dirname, '../next.config.js'),
          },
        },
      ],
    } as StorybookConfig;

    module.exports = config;
  `
  );
  tree.write(
    `plugin-apps/plugin/project.json`,
    JSON.stringify({
      name: 'plugin-app',
      root: 'plugin-apps/plugin',
      sourceRoot: 'plugin-apps/plugin/src',
      type: 'application',
      targets: {},
    })
  );
}
