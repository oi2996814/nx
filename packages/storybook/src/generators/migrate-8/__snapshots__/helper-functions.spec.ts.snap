// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper functions for the Storybook 8 migration generator getAllStorybookInfo and onlyShowGuide should return all info for all projects with Storybook 1`] = `
{
  "nextapp": {
    "configDir": "apps/nextapp/.storybook",
  },
  "nextapp-ts": {
    "configDir": "apps/nextapp-ts/.storybook",
  },
  "ngapp": {
    "configDir": "apps/ngapp/.storybook",
  },
  "ngapp-ts": {
    "configDir": "apps/ngapp-ts/.storybook",
  },
  "plugin-app": {
    "configDir": "plugin-apps/plugin/.storybook",
  },
  "rv1": {
    "configDir": "apps/rv1/.storybook",
  },
  "rv2-ts": {
    "configDir": "apps/rv2-ts/.storybook",
  },
  "rw1": {
    "configDir": "apps/rw1/.storybook",
  },
  "wv1": {
    "configDir": "apps/wv1/.storybook",
  },
  "ww1": {
    "configDir": "apps/ww1/.storybook",
  },
}
`;

exports[`Helper functions for the Storybook 8 migration generator logResult should create the summary file with the correct content 1`] = `
"# Storybook 8 Migration Summary

## Upgrade Storybook packages

The following command was ran to upgrade the Storybook packages:

\`\`\`bash
npx storybook@latest upgrade
\`\`\`

## The Storybook automigration scripts were ran


The following commands ran successfully and your Storybook configuration was successfully migrated to the latest version 7:


- \`npx storybook@latest automigrate --config-dir apps/nextapp/.storybook\`

- \`npx storybook@latest automigrate --config-dir apps/rv1/.storybook\`


Please make sure to check the results yourself and make sure that everything is working as expected.



### Some migrations failed

The following commands failed and your Storybook configuration for these projects was not
migrated to the latest version 7:


- \`npx storybook@latest automigrate --config-dir apps/rv2-ts/.storybook\`

- \`npx storybook@latest automigrate --config-dir apps/rw1/.storybook\`


You can run these commands again, manually, and follow the instructions in the
output of these commands to migrate your Storybook configuration to the latest version 8.


Also, we may have missed something. Please make sure to check the logs of the Storybook CLI commands that were run, and look for
the \`❌ Failed trying to evaluate\` message or \`❌ The migration failed to update\` message. This will indicate if a command was
unsuccessful, and will help you run the migration again, manually.

## Next steps

You can make sure everything is working as expected by trying
to build or serve your Storybook as you normally would.

\`\`\`bash
npx nx build-storybook project-name
\`\`\`

\`\`\`bash
npx nx storybook project-name
\`\`\`

Please read the [Storybook 8.0.0 release article](https://storybook.js.org/blog/storybook-8/) and the
official [Storybook 8.0.0 migration guide](https://storybook.js.org/docs/react/migration-guide)
for more information.
"
`;
